### ISSUE TITLE
The final boss

### ISSUE THREAD

**Issue by VoidXH** [VoidXH/Cavern#201]
This is it. The best way to easily add real-time Atmos decoding to media players. No DirectShow, no plugins, easy to integrate. Create a named pipe host that runs in the background, optionally starting with the PC, that takes any incoming data, and render it as PCM with the current Cavern setup. Make a reference client implementation in C++, and start contributing it to some open source players.

**Comment by VoidXH** [comment]
Beta implementation done in 20a19483087ff94cd29b4921c644825a254ace61. Protocol finalized, media players could theoritically start implemented. Need to fix a single bug marked with TODO and add full file processing solution to reference implementation. Documentation is also WIP.

**Comment by VoidXH** [comment]
Beta implementation done in 20a19483087ff94cd29b4921c644825a254ace61. Protocol finalized, media players could theoretically start implementing it. Need to fix a single bug marked with TODO and add full file processing solution to reference implementation. Documentation is also WIP.

**Comment by VoidXH** [comment]
Bugs fixed and language handling added in 0c281c8d94ffbfeeb5792eb453d9cd313d0495f8. Remains:
- The window would look nice with channel meters and a green/yellow/red indicator if the server is running, waiting for playback, or not running.
- Documentation, containing all protocol bytes. It shall include that sending a single 0 byte triggers retrieving all rendered frames.
- Reference client shall actively wait for all rendered frames.

**Comment by VoidXH** [comment]
CavernPipe is finalized and documented in 9f8ddc828d4ed719eebdf0413065e8b825d52e6f. To close this issue, implementation in any player is required, then I release CavernPipe bundled with that player on the website, with a single installer, that installs CavernPipe to be automatically launched with Windows.

**Comment by ThreeDeeJay** [comment]
So this requires the player to be modified? Or could we eventually just [passthrough Atmos](https://github.com/VoidXH/Cavern/issues/61) via a virtual audio device (or loopback a real S-PDIF device) into Cavern? Perhaps @dechamps would know if/how that'd be possible. 👀 

sView would be a great open source player to implement this in (unless it could eventually [decode and play LAF](https://github.com/gkv311/sview/issues/132) or perhaps even Atmos natively), though I think PotPlayer is hands down the best free player on Windows feature-wise, especially for [3D content](https://github.com/ThreeDeeJay/PotPlayer3D/releases/latest), though it's closed-source. 🤔

**Comment by dechamps** [comment]
> could we eventually just [passthrough Atmos](https://github.com/VoidXH/Cavern/issues/61) via a virtual audio device (or loopback a real S-PDIF device) into Cavern?

This would require one to write a Windows virtual audio driver which exposes Atmos bitstreaming capability and then hands the stream back to Cavern in userspace. I think that's technically doable, but writing Windows audio drivers is definitely not for the faint of heart. Modifying players would be much easier.

**Comment by EaryChow** [comment]
> This would require one to write a Windows virtual audio driver which exposes Atmos bitstreaming capability and then hands the stream back to Cavern in userspace.

There are freely available solutions like Voicemeeter, VB Cable, VB Matrix, etc. Though these are all from VB audio, but they are free to use. Voicemeeter and VB Matrix seems to only support 8 channel audio at most, but VB Cable has 16 channel audio support. I can imagine a chain where we take in bitstream into Voicemeeter's virtual microphone, and then Cavern picks up the mic signal and decode, output to to VB Cable, and then I can use VB Cable to send the decoded PCM signals to the binaural renderer APL Virtuoso that I have been using. Cavern only needs to be able to pick up a microphone signal since there are third party solutions.

**Comment by dechamps** [comment]
> There are freely available solutions like Voicemeeter, VB Cable, VB Matrix, etc.

I would be very surprised if any of these advertise Atmos bitstreaming support of their virtual audio outputs. I'd expect you'd need to ask for that as a new feature.

**Comment by EaryChow** [comment]
> I would be very surprised if any of these advertise Atmos bitstreaming support on their virtual audio devices. I'd expect you'd need to ask for that as a new feature. 

Good point, just did some search, it seems in some forum posts they said they only support PCM streams. 

But having a virtual input option in Cavern is still a nice move to prepare for the potential future. Cavern doesn't have to ship a virtual cable with it, having a slot is better than nothing, I guess.  I think it would also make it easier to make the feature request to the cables.

**Comment by VoidXH** [comment]
One thing to note regarding virtual audio devices, is how bitstreams are actually handled over SPDIF or HDMI. While there is metadata over HDMI regarding the codec, it's still SPDIF-compatible: set the stereo sample rate to 192 kHz, and just push the AC-3 bytes through. Cavern can definitely pick this up and play it back.

**However**, as regular bitstreams, this will make you lose volume control in-player, will add a detection delay, always has to run a microphone, which will scare privacy-concerned users, etc. CavernPipe has none of these issues.

**Comment by ThreeDeeJay** [comment]
Dang, it's sad people don't seem to appreciate such a neat feature.
If you can figure out the LAVFilters code, maybe a pull request (even if it's just a sloppy proof of concept) could make the devs come out of the woodwork and help carry out the merge. I would definitely use a Kodi fork though, especially if it works with the streaming service addons that support Atmos tracks.

But if none of the above are possible, what would be the cons of an independent DirectShow plugin?
AFAIK Kodi, MPC and others like PotPlayer could load it without having to jump through the hops of bureaucracy of merging the CavernPipe compatibility changes. and if so, would it be feasible to make the plugin standalone/self-contained so that it just loads a tray icon we can click to configure the output mode? 🤔

**Comment by VoidXH** [comment]
> what would be the cons of an independent DirectShow plugin?

Barely anyone knows how to even begin a project with it, I asked multiple audio devs. It's a 20+ year old legacy Windows SDK dark magic that somehow still works. Modern .NET isn't even COM-compatible, so it's not native, but CavernPipe-only.

**Comment by ThreeDeeJay** [comment]
Dang I hadn't realized how ancient this thing is. I just found this
https://github.com/EnsignPayton/DshowTestFilters
https://github.com/CSIR-RTVC/DirectShowFilterPack

But isn't LAVFilters basically a DirectShow plugin/collection though? like if it works as an addon for LAVFilters it I'd imagine it'd be able to work on its own. But yeah, working with such ancient software must be a PITA so it's probably better to piggy-back on LAVFilters anyway.

**Comment by VoidXH** [comment]
LAV is DS, yes. Multiple steps were done towards DS compatibility like #85, but all have failed for some reason or the other.

**Comment by ThreeDeeJay** [comment]
Any glimpse of hope for this yet?
I'm afraid the PotPlayer dev (or at least whoever handles their email responses) don't like the idea, despite me trying my best to make a case for it  💀  
<img width="197" height="77" alt="Image" src="https://github.com/user-attachments/assets/f4b55489-ae55-43d1-aab3-fc6191e7b487" />

**Comment by VoidXH** [comment]
LAV is supportive of the idea. The dev is very busy though.

**Comment by Glider95** [comment]
For test purpose, I wanted to cascade:
Kodi -> Snapserver -> Snapclient -> Cavern.

I thought I was smart but now I understand what you want to get the (dolby) bitstream out of Kodi.  My idea is just getting  PCM/FLAC from Snapserver....
